p2panda-rs/src/storage_provider/traits/storage_provider.rs

```diff
@@ -0,0 +1,512 @@
+            .await?;
+
+        // Determine backlink and skiplink hashes for the next entry. To do this we need the latest
+        // entry in this log
+        let entry_latest = self.latest_entry(params.author(), &log).await?;
+
+        match entry_latest.clone() {
+            // An entry was found which serves as the backlink for the upcoming entry
+            Some(entry_backlink) => {
+                let entry_latest = entry_latest.unwrap();
+                let entry_hash_backlink = entry_backlink.hash();
+                // Determine skiplink ("lipmaa"-link) entry in this log
+                let entry_hash_skiplink = self.determine_skiplink(&entry_latest).await?;
+
+                Ok(Self::EntryArgsResponse::new(
+                    Some(entry_hash_backlink.clone()),
+                    entry_hash_skiplink,
+                    entry_latest.seq_num().clone().next().unwrap(),
+                    entry_latest.log_id(),
+                ))
+            }
+            // No entry was given yet, we can assume this is the beginning of the log
+            None => Ok(Self::EntryArgsResponse::new(
+                None,
+                None,
+                SeqNum::default(),
+                log,
+            )),
+        }
+    }
+
+    /// Stores an author's Bamboo entry with operation payload in database after validating it.
+    async fn publish_entry(
+        &self,
+        params: &Self::PublishEntryRequest,
+    ) -> Result<Self::PublishEntryResponse, Box<dyn std::error::Error>> {
+        // Create a storage entry.
+        let entry = StorageEntry::new(params.entry_signed(), params.operation_encoded())?;
+        // Validate the entry (this also maybe happened in the above constructor)
+        entry.validate().map_err(|error| format!("{:?}", error))?;
+
+        // Every operation refers to a document we need to determine. A document is identified by the
+        // hash of its first `CREATE` operation, it is the root operation of every document graph
+        let document_id = if entry.operation().is_create() {
+            // This is easy: We just use the entry hash directly to determine the document id
+            DocumentId::new(entry.hash().into())
+        } else {
+            // For any other operations which followed after creation we need to either walk the operation
+            // graph back to its `CREATE` operation or more easily look up the database since we keep track
+            // of all log ids and documents there.
+            //
+            // We can determine the used document hash by looking at what we know about the previous
+            // entry in this author's log.
+            //
+            // @TODO: This currently looks at the backlink, in the future we want to use
+            // "previousOperation", since in a multi-writer setting there might be no backlink for
+            // update operations! See: https://github.com/p2panda/aquadoggo/issues/49
```

Comment:
What is stopping us?

Reply:
I don't think there's anything stopping us. I tried to change as little of the inner code in all these existing methods as possible (they all came over from `aquadoggo`) as I wanted to limit the scope of this PR to the restructuring of our API into traits rather than the implementation details. Maybe i shouldn't have done that, but it seemed like a useful limit when i started.

Reply:
Ok... against my own advice, I've now implemented following `previous_operations` to find the doucment id instead of the backlink. I had to add some extra validation in `Operation` to do this.